5.8 + 16.8 Cannot resolve transitive project dependencies when mixing PackageReference and packages.config

## Details about Problem

NuGet product used (NuGet.exe | VS UI | Package Manager Console | dotnet.exe):

NuGet version (5.8):
VS version (if appropriate): 16.8.x

Worked before? If so, with which NuGet version
Not entirely sure, but we did not experience an issue before installing 5.8 + 16.8.X
NuGet version (5.7):
VS version (if appropriate): 16.7.x

## Detailed repro steps so we can see the same problem

Given a project dependency chain::

```
TestProject (netcoreapp3.1) 
╚ MvcWebApplication (net461, packages.config old-style csproj) 
  ╚ DependencyA (netstandard2) 
    ╚ TransientProjectB (netstandard2)
```

The TestProject use classes from TransientProjectB directly in a test.

This fails with 
>D:\a\mixed-repro\mixed-repro\TestProject\UnitTest1.cs(3,7): error CS0246: The type or namespace name 'TransientProjectB' could not be found (are you missing a using directive or an assembly reference?) [D:\a\mixed-repro\mixed-repro\TestProject\TestProject.csproj]

The solution builds _fine_ in the IDE (Visual Studio 2019 16.8.X), but not via msbuild on the command line (and hence, not on our build servers).

It's easily enough fixed by adding a direct reference instead of keeping it transient, but was this breaking change on purpose? Why did it work in 16.7 whilst not in 16.8?

### Sample Project
Repo: https://github.com/icenorge/mixed-repro
Build script: https://github.com/icenorge/mixed-repro/blob/master/.github/workflows/MSBuild.yml
Failing build: https://github.com/icenorge/mixed-repro/runs/1453122655?check_suite_focus=true

--- a/test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs
+++ b/test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs
@@ -10187,7 +10187,7 @@ namespace NuGet.CommandLine.Test
                 // B -> packages.config
                 Util.CreateFile(Path.GetDirectoryName(projectB.ProjectPath), "packages.config",
 @"<packages>
-  <package id=""X"" version=""1.1.0"" targetFramework=""NETStandard1.5"" />
+  <package id=""X"" version=""1.1.0"" targetFramework=""net472"" />
 </packages>");
 
                 solution.Projects.Add(projectA);

--- a/test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs
+++ b/test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs
@@ -959,10 +959,15 @@ namespace NuGet.CommandLine.Test
                 // Assert
                 Assert.True(0 == r.Item1, r.Item2 + " " + r.Item3);
 
-                // Assert
+                // Assert correct projects were restored.
                 Assert.True(File.Exists(projectA.AssetsFileOutputPath));
                 Assert.False(File.Exists(projectB.AssetsFileOutputPath));
                 Assert.True(File.Exists(projectC.AssetsFileOutputPath));
+
+                // Assert transitivity is applied across non PackageReference projects.
+                var ridlessTarget = projectA.AssetsFile.Targets.Where(e => string.IsNullOrEmpty(e.RuntimeIdentifier)).Single();
+                ridlessTarget.Libraries.Should().Contain(e => e.Type == "project" && e.Name == projectB.ProjectName);
+                ridlessTarget.Libraries.Should().Contain(e => e.Type == "project" && e.Name == projectC.ProjectName);
             }
         }
 
@@ -10132,6 +10137,96 @@ namespace NuGet.CommandLine.Test
                     .SelectMany(g => g.TransitiveDependencies.Select(t => $"{g.FrameworkName}_{t.LibraryRange.Name}.{t.LibraryRange.VersionRange.OriginalString}")).ToList();
 
                 Assert.Equal(0, centralfileDependencyGroups.Count);
+            }
+        }
+
+        [Fact]
+        public async Task RestorePackageReference_WithPackagesConfigProjectReference_IncludesTransitivePackageReferenceProjects()
+        {
+            // Arrange
+            using (var pathContext = new SimpleTestPathContext())
+            {
+                // Set up solution, project, and packages
+                var solution = new SimpleTestSolutionContext(pathContext.SolutionRoot);
+
+                var projectA = SimpleTestProjectContext.CreateNETCore(
+                    "a",
+                    pathContext.SolutionRoot,
+                    NuGetFramework.Parse("NETFramework4.7.2"));
+
+                var projectB = SimpleTestProjectContext.CreateNonNuGet(
+                    "b",
+                    pathContext.SolutionRoot,
+                    NuGetFramework.Parse("NETFramework4.7.2"));
+
+                var projectC = SimpleTestProjectContext.CreateNETCore(
+                    "c",
+                    pathContext.SolutionRoot,
+                    NuGetFramework.Parse("NETFramework4.7.2"));
+
+                var packageX100 = new SimpleTestPackageContext("X", "1.0.0");
+                var packageX110 = new SimpleTestPackageContext("X", "1.1.0");
+                var packageY = new SimpleTestPackageContext("Y", "1.0.0");
+
+                projectA.AddPackageToAllFrameworks(packageX100);
+                projectC.AddPackageToAllFrameworks(packageY);
+
+                await SimpleTestPackageUtility.CreateFolderFeedV3Async(
+                    pathContext.PackageSource,
+                    PackageSaveMode.Defaultv3,
+                    packageX100,
+                    packageX110,
+                    packageY);
+
+                // A -> B
+                projectA.AddProjectToAllFrameworks(projectB);
+
+                // B -> C
+                projectB.AddProjectToAllFrameworks(projectC);
+
+                // B -> packages.config
+                Util.CreateFile(Path.GetDirectoryName(projectB.ProjectPath), "packages.config",
+@"<packages>
+  <package id=""X"" version=""1.1.0"" targetFramework=""NETStandard1.5"" />
+</packages>");
+
+                solution.Projects.Add(projectA);
+                solution.Projects.Add(projectB);
+                solution.Projects.Add(projectC);
+                solution.Create(pathContext.SolutionRoot);
+
+                // Act
+                var nugetexe = Util.GetNuGetExePath();
+
+                var args = new string[] {
+                    "restore",
+                    projectA.ProjectPath,
+                    "-Verbosity",
+                    "detailed",
+                    "-Recursive"
+                };
+
+                // Act
+                var r = CommandRunner.Run(
+                    nugetexe,
+                    pathContext.WorkingDirectory.Path,
+                    string.Join(" ", args),
+                    waitForExit: true);
+
+                // Assert
+                Assert.True(0 == r.Item1, r.Item2 + " " + r.Item3);
+
+                // Assert correct projects were restored.
+                Assert.True(File.Exists(projectA.AssetsFileOutputPath));
+                Assert.False(File.Exists(projectB.AssetsFileOutputPath));
+                Assert.True(File.Exists(projectC.AssetsFileOutputPath));
+
+                // Assert transitivity is applied across non PackageReference projects.
+                var ridlessTarget = projectA.AssetsFile.Targets.Where(e => string.IsNullOrEmpty(e.RuntimeIdentifier)).Single();
+                ridlessTarget.Libraries.Should().Contain(e => e.Type == "project" && e.Name == projectB.ProjectName);
+                ridlessTarget.Libraries.Should().Contain(e => e.Type == "project" && e.Name == projectC.ProjectName);
+                ridlessTarget.Libraries.Should().Contain(e => e.Name == "X");
+                ridlessTarget.Libraries.Should().Contain(e => e.Name == "Y");
             }
         }
 

